plugin/fallback/upstream_utils.go

```diff
@@ -0,0 +1,50 @@
+// Package fallback implements a fallback plugin for CoreDNS
+package fallback
+
+import (
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/plugin/proxy"
+)
+
+// upstreamMapper manages map of rcode to upstream
+type upstreamMapper interface {
+	Add(int, proxy.Upstream) proxy.Upstream
+	Get(int) (proxy.Upstream, bool)
+}
+
+// fallbackUpstreamMapper mplements the upstreamMapper interface
+// Used by the fallback plugin
+type fallbackUpstreamMapper struct {
+	upstreamMap map[int]proxy.Upstream
+}
+
+func (f *fallbackUpstreamMapper) Add(rcode int, u proxy.Upstream) (oldU proxy.Upstream) {
+	var ok bool
+	if oldU, ok = f.upstreamMap[rcode]; !ok {
+		f.upstreamMap[rcode] = u
+	}
+
+	return
+}
+
+func (f *fallbackUpstreamMapper) Get(rcode int) (u proxy.Upstream, ok bool) {
+	u, ok = f.upstreamMap[rcode]
+	return
+}
+
+func newFallbackUpstreamMapper() upstreamMapper {
+	return &fallbackUpstreamMapper{upstreamMap: make(map[int]proxy.Upstream)}
+}
+
+// proxyCreator creates a proxy with the specified upstream
+type proxyCreator interface {
+	Create(trace plugin.Handler, upstream proxy.Upstream) plugin.Handler
+}
+
+// fallbackProxyCreator implements the proxyCreator interface
+// Used by the fallback plugin to create proxy using specified for upstream
+type fallbackProxyCreator struct{}
+
+func (f fallbackProxyCreator) Create(trace plugin.Handler, upstream proxy.Upstream) plugin.Handler {
+	return &proxy.Proxy{Trace: trace, Upstreams: &[]proxy.Upstream{upstream}}
```

Comment:
Same here - simpler to just have this one line of code in the setup?

Reply:
Please let me clarify. The `proxyCreator` interface is introduced so that I can write test to verify whether proxy is created using the expected upstream. This is the implementation that is actually used in the fallback plugin. There is a fake one used by unit test.

Not sure if this answer you question...

Reply:
Yeah, I figured it was for unit tests but somehow I missed that it was in fact used. OK.